crates/compilers/src/project_util/mod.rs

```diff
@@ -31,14 +31,20 @@ pub mod mock;
 /// A [`Project`] wrapper that lives in a new temporary directory
 ///
 /// Once `TempProject` is dropped, the temp dir is automatically removed, see [`TempDir::drop()`]
-pub struct TempProject<C: Compiler = MultiCompiler, T: ArtifactOutput = ConfigurableArtifacts> {
+pub struct TempProject<
+    C: Compiler = MultiCompiler,
+    T: ArtifactOutput<CompilerContract = C::CompilerContract> = ConfigurableArtifacts,
+> {
     /// temporary workspace root
     _root: TempDir,
     /// actual project workspace with the `root` tempdir as its root
     inner: Project<C, T>,
 }
 
-impl<T: ArtifactOutput + Default> TempProject<MultiCompiler, T> {
+impl<
+        T: ArtifactOutput<CompilerContract = <MultiCompiler as Compiler>::CompilerContract> + Default,
+    > TempProject<MultiCompiler, T>
```

Comment:
Perhaps we could think about a type alias for this.

Reply:
Yes it might make sense, want to see what foundry team says about the abstraction first because they probably have their own convention on how to handle types and names